### PR TITLE
First version of turbo cache

### PR DESCRIPTION
**Pull request by andectionsharechat**
## Context
It's a [fast cache](https://github.com/VictoriaMetrics/fastcache) fork. It reduces contention by reducing the time spent under the mutex during writing. 

## Fast cache implementation
Fast cache implementation is pretty simple. The key idea is that all data is stored in an array of chunks of predefined size. All items are written sequentially by this structure:` {encoded key&value lengths, key, value}`.
There is an index (just a map) which helps to find the item index in the chunks. All operations are executed under Read and Write Mutex. All keys are spread by 512 buckets, which is calculated by `hash(key) % bucket count`

image

## Turbo cache: overview
Turbocache extends fast cache by several changes
1.  On writing to turbo cache, all new items are put into a channel and processed in the background gorutine. The number of goroutine equals the number of buckets.
2. The goroutine writes consequences for new items to specific flush chunks. It doesn't need mutex, as all changes happen in the same thread. Periodically, the turbo cache flushes all items to the main chunks
3. Turbo cache uses specific **index** for reading items from the flush chunks as well
image

## Turbo cache: configuration

```
type Config struct {
	//max bytes for storing keys in chunks
	maxBytes int
	// flush intervals to writing keys to the chunks
	flushIntervalMillis int64
	//max batch size for writing in the chunks. batch size 1 make turbo cache to sync cache
	maxWriteBatch int
	//count of the accumulating buffers (chunks) before flush. Every flush chunk has 64KB
	flushChunkCount int
}

func NewConfig(maxBytes int, flushInterval int64, maxWriteBatch int, flushChunks int) *Config {
	.....
}
```

### Principles of thread synchronisation during reading from the index
1. All shared structures are arrays or splices with fixed-size
2. On processing new items, all data are appended 
3. All mutations happen only in one goroutine
4. Cleaning flush chunks && index happens under mutex-free lock

### On new key
1. Search by index for duplication. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L53)
2. Write keyValue in flush chunk. It's append-only to an already allocated array, with no memory recopying. [Pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L66)
3. Update value to chunk with memory barrier. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L69)
4. Add a new value to the index. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L76).  For writing, we first update flushChunkIndex, index inside the array and then only hash (during reading, everything is checked in the opposite order).
5. Memory barrier for index. [Code pointer  ](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L85)

###  On batch and cleaning flush chunks && index
this operation needs to be synchronised with reading from the index. The key idea is 2 step locking. 
1. Set flushing flag for preventing new readings from the index. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L139)
2. Take a spin lock for reading. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L140)
3. Cleaning index and flush chunks. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L142)
4. Release the flushing flag. 
5. Release the spinlock.

###  On index reading
Reading from index happens only on cache miss
1. Check if the flushing flag is set.[ Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L45)
2. Read index with memory barrier. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L45)
3. Search for the key hash in the index. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L52)
4. On success, try to acquire the spinlock. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L53)
5. On success, find an index to flush the chunk and read data from it. [Code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L58)

## Turbo cache: writing implementation
1. Determine bucket and write to the channel. [code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache.go#L474)
7. How a background goroutine processes new key. [code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L46)
8. How chunks are updated periodically. [code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush.go#L111)

## Turbo cache: reading implementation
1. reading from the index before acquiring mutex. [code pointer](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_index.go#L39)

## Turbo cache: testing
[Tests](https://github.com/ShareChat/turbo-cache/blob/v1/fastcache_flush_test.go)
1. Tests over public API + tests over updating keys without gourutine processes. All tests with different batch size
All tests are running with --race argument

## Turbo cache: benchmarking
1. During developing, there were benchmarks with comparative CPU time over mutex
7. It misses the comparison between sync fast cache and async cache

## Turbo cache: QA
What's memory overhead?
1. it's an overhead with flush chunks + index. Overhead: 512 (bucket count) *  64KB * flushChunkCount + maxBatchSize * 84B. It's about 33MB for flushChunkCount=1, maxBatchSize=128

### CODE REVIEW

File: bigcache.go [hunk 1]
```diff
@@ -43,26 +42,30 @@ func (c *Cache) SetBig(k, v []byte) {
...
+		bytes := subkeys[startSubkeyIndex:]
+		startSubkeyIndex = len(subkeys)
+		c.Set(bytes, v[startValueIndex:endValueIndex])
 	}
-
 	// Write metavalue, which consists of valueHash and valueLen.
-	subkey.B = marshalUint64(subkey.B[:0], valueHash)
-	subkey.B = marshalUint64(subkey.B, uint64(valueLen))
-	c.Set(k, subkey.B)
-	putSubkeyBuf(subkey)
```
Comment (andectionsharechat) [inline bigcache.go:65]: Removed the pulling object as we don't control lifecycle here anymore

File: fastcache_flush.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	b.gen = b.gen[:0]
+}
+
+func (b *bucket) randomDelay(maxDelayMillis int64) {
+	jitterDelay := rand.Int63() % (maxDelayMillis * 1000)
+	time.Sleep(time.Duration(jitterDelay) * time.Microsecond)
+}
+
+func (f *flusher) incrementIndexes(kvLength uint64) (idxNew uint64, newChunk bool) {
+	idxNew = f.idx + kvLength
```
Comment (andectionsharechat) [inline fastcache_flush.go]: it's really hard to read, I will think how to rethink this

File: fastcache_flush_index.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
+// Package fastcache implements fast in-memory cache.
+//
+// The package has been extracted from https://victoriametrics.com/
+package turbocache
+
+import "sync/atomic"
+
+type flushChunkIndexItem struct {
+	//flush chunk number
+	flushChunk [7]int32
```
Comment (andectionsharechat) [inline fastcache_flush_index.go]: Chunk and flushChunk seems like bad terminology 🤔 it needs to be changed

File: fastcache_batch_timing_test.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
+package turbocache
+
+import (
+	"github.com/cespare/xxhash/v2"
+	"runtime/debug"
+	"testing"
+)
+
+func BenchmarkBatchSetMostOptimised(b *testing.B) {
```
Comment (andectionsharechat) [inline fastcache_batch_timing_test.go]: will create benchmark fast cache VS turbo cache

File: fastcache_batch_timing_test.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	c := New(newCacheConfigWithDefaultParams(12000 * items))
+	defer c.Reset()
+	k := []byte("\x00\x00\x00\x00")
+	v := []byte("xyza")
+
+	buffer := make([]flushChunk, 2)
+	for i := 0; i < 2; i++ {
+		for j := 0; j < buffSize/2; j++ {
+			k[0] = byte(j)
+			//kvLenBuf := makeKvLenBuf(k, v)
```
Comment (isburmistrov) [inline fastcache_batch_timing_test.go]: what is this comment?

File: fastcache_batch_timing_test.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		}
+
+	}
+	b.ReportAllocs()
+	debug.SetGCPercent(-1)
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			for i := 0; i < items; i++ {
+				//		c.buckets[i%bucketsCount].setBatch(buffer, true, 1, 1)
```
Comment (isburmistrov) [inline fastcache_batch_timing_test.go]: ????

File: fastcache_batch_timing_test.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+			}
+		}
+	})
+	debug.SetGCPercent(100)
+}
+
+/*
+func BenchmarkBatchSet(b *testing.B) {
```
Comment (isburmistrov) [inline fastcache_batch_timing_test.go]: hmmm

File: fastcache_flush.go [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+		kvLength := uint64(4) + uint64(len(i.K)) + uint64(len(i.V))
+		idxNew, newChunk := f.incrementIndexes(kvLength)
+		if newChunk {
+			f.currentFlushChunkId++
+		}
+		if f.currentFlushChunkId <= int32(len(f.chunks)) {
+			flushChunk := &f.chunks[f.currentFlushChunkId]
+
+			lenBuf := makeKvLenBuf(i.K, i.V)
+			//f.spinlock.Lock()
```
Comment (isburmistrov) [inline fastcache_flush.go]: what do these Lock / Unlock comments mean?

File: fastcache.go [hunk 9]
```diff
@@ -248,59 +304,189 @@ func (b *bucket) Init(maxBytes uint64) {
...
+	if sync {
+		b.syncSet(k, v, h)
+	} else {
+		iv := getQueuedStruct(k, v, h)
+
+		select {
+		case b.setBuf <- iv:
+			return
+		default:
+			atomic.AddUint64(&b.dropsInQueue, 1)
```
Comment (isburmistrov) [inline fastcache.go]: don't understand what does it mean. Does it mean that we drop writes in some cases?

File: fastcache.go [hunk 10]
```diff
@@ -248,59 +304,189 @@ func (b *bucket) Init(maxBytes uint64) {
...
+func (b *bucket) set(kv []byte, h uint64, idx uint64, chunkIdx uint64, needClean bool) (newNeedClean bool, newIdx uint64, newChunkId uint64) {
+	chunks := b.chunks
+	idxNew := idx + uint64(len(kv))
+	chunkIdxNew := idxNew / chunkSize
+	if chunkIdxNew > chunkIdx {
+		if chunkIdxNew >= uint64(len(chunks)) {
+			idx = 0
+			idxNew = uint64(len(kv))
+			chunkIdx = 0
+			b.gen++
```
Comment (isburmistrov) [inline fastcache.go]: IMO this needs some separate function with a clear name - it's hard to figure out what's going on with this `gen` thing. Or this code is forked? Shows up like a new...

File: fastcache_flush.go [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+	b.idx.Store(f.idx)
+	b.gen = f.gen
+	if f.needClean {
+		b.cleanLocked(f.idx)
+	}
+}
+
+func (b *bucket) cleanFlusher(f *flusher) {
+	index := b.flusher.index.Load().([]flushChunkIndexItem)
+	f.flushed.Store(true)
```
Comment (isburmistrov) [inline fastcache_flush.go]: I don't quite get this logic. So we take `flusher.index` first and then set `f.flushed` to `true`. But it means that we have race condition with `tryFindInFlushIndex`? That function is reading-only, but here we do some non-atomic stuff like 

```
if atomic.LoadUint64(&index[i].h[j]) != 0 {
				atomic.StoreUint64(&index[i].currentIdx[j], 0)
				atomic.StoreInt32(&index[i].flushChunk[j], 0)
				atomic.StoreUint64(&index[i].h[j], 0)
```

so I guess that function can be pretty much messed up

File: fastcache_flush.go [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+			for j := range index[indexId].h {
+				if atomic.LoadUint64(&index[indexId].h[j]) == 0 {
+					atomic.StoreUint64(&index[indexId].currentIdx[j], flushChunk.size)
+					atomic.StoreInt32(&index[indexId].flushChunk[j], f.currentFlushChunkId)
+					atomic.StoreUint64(&index[indexId].h[j], i.h)
+					break
+				}
+			}
+
+			b.flusher.index.Store(index)
```
Comment (isburmistrov) [inline fastcache_flush.go]: Don't understand. So we get `index` as `b.flusher.index.Load().([]flushChunkIndexItem)`, and here we `Store` it back. But don't we modify the existing slice anyway? Which actually means that the content of `index` may be inconsistent during these non-atomic calls

```
if atomic.LoadUint64(&index[indexId].h[j]) == 0 {
					atomic.StoreUint64(&index[indexId].currentIdx[j], flushChunk.size)
					atomic.StoreInt32(&index[indexId].flushChunk[j], f.currentFlushChunkId)
					atomic.StoreUint64(&index[indexId].h[j], i.h)
					break
				}
```

?

**Review by isburmistrov — COMMENTED** [review]
First attempt. Don't understand a few things - asked inline.

File: fastcache_batch_timing_test.go (hunk 5 above)
Comment (andectionsharechat) [reply fastcache_batch_timing_test.go]: it's basically just debug code. I used it for comparation different implemetation of writing under the mutext. the whole file is not relevant, want to rewrite it

File: fastcache_flush.go (hunk 8 above)
Comment (andectionsharechat) [reply fastcache_flush.go]: removed

File: fastcache_flush.go (hunk 12 above)
Comment (andectionsharechat) [reply fastcache_flush.go]: I will write it with details later. The key idea is:
1. we have chunk with ready to write data
2. we have index above it

during writing
1. add data to the chunk, just append bytes in array without recopying the array in a heap
2. append data in index (index item contains several `arrays`, not slices). the order makes sense here, firstly we add index to flush chunk, than hash of the key.

During **flushing** we use spin lock and clean index, than clean flush chunk itself

During reading from the index with firstly check hash in index, than we read bytes from so-called flush chunk. reading from flush chunks are executed under the spin lock.

Also I had a flag  about flushing and the read process doesnt acquire for the spin lock in this case.

So, basically we most of the time have append only writing, for reading bytes && cleaning the index with use spin lock

Comment (andectionsharechat) [reply fastcache_flush.go]: it's relevant to your second comment as well. I think I need to write it down in description as well with some visualisation

File: internal/spinlock/rwmutext.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
+package spinlock
```
Comment (isburmistrov) [inline internal/spinlock/rwmutext.go:1]: what's the difference with [sync.RWMutex](https://pkg.go.dev/sync#RWMutex) by the way?

Comment (andectionsharechat) [reply internal/spinlock/rwmutext.go:1]: the original uses os semaphore
